ARROW-3144: [C++/Python] Move "dictionary" member from DictionaryType to ArrayData to allow for variable dictionaries

wesm opened the pull request:
This patch moves the dictionary member out of DictionaryType to a new
member on the internal ArrayData structure. As a result, serializing
and deserializing schemas requires only a single IPC message, and
schemas have no knowledge of what the dictionary values are.

The objective of this change is to correct a long-standing Arrow C++
design problem with dictionary-encoded arrays where the dictionary
values must be known at schema construction time. This has plagued us
all over the codebase:

* In reading Parquet files, reading directly to DictionaryArray is not
  simple because each row group may have a different dictionary
* In IPC streams, delta dictionaries (not yet implemented) would
  invalidate the pre-existing schema, causing subsequent RecordBatch
  objects to be incompatible
* In Arrow Flight, schema negotiation requires the dictionaries to be
  sent, having possibly unbounded size.
* Not possible to have different dictionaries in a ChunkedArray
* In CSV files, converting columns to dictionary in parallel would
  require an expensive type unification

The summary of what can be learned from this is: do not put data in
type objects, only metadata. Dictionaries are data, not metadata.

There are a number of unavoidable API changes (straightforward for
library users to fix) but otherwise no functional difference in the
library.

As you can see the change is quite complex as significant parts of IPC
read/write, JSON integration testing, and Flight needed to be reworked
to alter the control flow around schema resolution and handling the
first record batch.

Key APIs changed

* `DictionaryType` constructor requires a `DataType` for the
  dictionary value type instead of the dictionary itself. The
  `dictionary` factory method is correspondingly changed. The
  `dictionary` accessor method on `DictionaryType` is replaced with
  `value_type`.
* `DictionaryArray` constructor and `DictionaryArray::FromArrays` must
  be passed the dictionary values as an additional argument.
* `DictionaryMemo` is exposed in the public API as it is now required
  for granular interactions with IPC messages with such functions as
  `ipc::ReadSchema` and `ipc::ReadRecordBatch`
* A `DictionaryMemo*` argument is added to several low-level public
  functions in `ipc/writer.h` and `ipc/reader.h`

Some other incidental changes:

* Because DictionaryType objects could be reused previous in Schemas, such dictionaries would be "deduplicated" in IPC messages in passing. This is no longer possible by the same trick, so dictionary reuse will have to be handled in a different way (I opened ARROW-5340 to investigate)

* As a result of this, an integration test that featured dictionary reuse has been changed to not reuse dictionaries. Technically this is a regression, but I didn't want to block the patch over it

* R is added to allow_failures in Travis CI for now

bkietz commented on cpp/src/arrow/type_traits.h:
```diff
@@ -292,7 +291,16 @@ struct TypeTraits<ExtensionType> {
 //
 
 template <typename T>
-using is_number = std::is_base_of<Number, T>;
```
If we're renaming this one, should we do the same for `is_signed_integer` etc?

wesm replied:
Probably. There's some other cleanup to do but not here, I will open a JIRA

bkietz commented on cpp/src/arrow/util/concatenate.cc:
```diff
@@ -210,7 +210,24 @@ class ConcatenateImpl {
-    return ConcatenateBuffers(Buffers(1, *fixed), pool_, &out_.buffers[1]);
+
+    // Two cases: all the dictionaries are the same, or unification is
+    // required
```
:+1:

bkietz commented on cpp/src/arrow/type.h:
```diff
@@ -778,48 +791,54 @@ class ARROW_EXPORT IntervalType : public FixedWidthType {
-/// Concrete type class for dictionary data
+/// \brief Dictionary-encoded value type with data-dependent
+/// dictionary
 class ARROW_EXPORT DictionaryType : public FixedWidthType {
```
After this refactor DictionaryType looks more like a nested type.

wesm replied:
It's a synthetic construct in C++ since there is no Dictionary type in the protocol metadata...

wesm commented on cpp/src/arrow/array.h:
```diff
@@ -206,6 +208,10 @@ struct ARROW_EXPORT ArrayData {
+
+  // The dictionary for this Array, if any. Only used for dictionary
+  // type
+  std::shared_ptr<Array> dictionary;
```
This was discussed on the mailing list. I agree with the others (Antoine / Micah) that having an explicit dictionary field is more clear. I added a benchmark to assess if it causes meaningful overhead which it does not seem to.

bkietz replied:
Relevant mailing list entry: https://lists.apache.org/thread.html/987a9f9a36f491e49d71ddcf9ce91f03420210ec30dd2c24895b2332@%3Cdev.arrow.apache.org%3E

wesm commented on cpp/src/arrow/array.cc:
```diff
@@ -765,20 +763,30 @@ static Status TransposeDictIndices(MemoryPool* pool, const ArrayData& in_data,
                                   std::shared_ptr<Array>* out) const {
   DCHECK_EQ(type->id(), Type::DICTIONARY);
   const auto& out_dict_type = checked_cast<const DictionaryType&>(*type);
 
```
Possibly, but out of scope for this patch

wesm commented on cpp/src/arrow/compute/kernels/cast.cc:
```diff
@@ -746,213 +743,159 @@ class FromNullCastKernel : public CastKernelBase {
       default:
         ctx->SetStatus(
-            Status::Invalid("Invalid index type: ", indices.type()->ToString()));
+            Status::Invalid("Invalid index type: ", type.index_type()->ToString()));
```
I prefer to leave the semantics here as unchanged as possible from master

wesm commented on cpp/src/arrow/array-dict-test.cc:
```diff
@@ -749,47 +773,48 @@ TEST(TestDecimalDictionaryBuilder, DoubleTableSize) {
 // ----------------------------------------------------------------------
 // DictionaryArray tests
 
 TEST(TestDictionary, Basics) {
```
Yes, will do

pitrou commented on cpp/src/arrow/python/flight.cc:
```diff
@@ -196,9 +200,14 @@ PyGeneratorFlightDataStream::PyGeneratorFlightDataStream(
+std::shared_ptr<Schema> PyGeneratorFlightDataStream::schema() { return schema_; }
+
+Status PyGeneratorFlightDataStream::GetSchemaPayload(FlightPayload* payload) {
+  return ipc::internal::GetSchemaPayload(*schema_, &dictionary_memo_,
```
I'm not sure I understand. This seems to populate a `DictionaryMemo` but it's not used afterwards?

pitrou replied:
We don't have a test for dictionary arrays in `test_flight.py`...

lihalite replied:
Last I tried, cross-language Flight with dictionaries still didn't work, or even from C++ to C++, so it wouldn't have worked before in Python. https://issues.apache.org/jira/browse/ARROW-5143

pitrou replied:
Ah... They work with `DoGet` but perhaps not with `DoPut` then?

pitrou replied:
@lihalite I'm fixing dict transfer with DoPut as part of ARROW-5113. It may produce conflicts for both you and @wesm  :-)

wesm replied:
OK, you may want to abort the dict transfer work until this is merged

pitrou commented on cpp/src/arrow/array.cc:
```diff
@@ -799,9 +807,8 @@ Status DictionaryArray::Transpose(MemoryPool* pool, const std::shared_ptr<DataTy
+#undef TRANSPOSE_IN_OUT_CASE
 }
 
 // ----------------------------------------------------------------------
```
I think `ValidateArray` should check `data()->dictionary` for dictionary arrays.

pitrou commented on cpp/src/arrow/flight/server.h:
```diff
@@ -45,9 +46,11 @@ class ARROW_EXPORT FlightDataStream {
   virtual std::shared_ptr<Schema> schema() = 0;
 
+  /// \brief Compute FlightPayload containing serialized RecordBatch schema
+  virtual Status GetSchemaPayload(FlightPayload* payload) = 0;
```
(or is it @lihalite ?)

lihalite replied:
Either will ping me!

I think in this case, it's OK, in that Flight explicitly states we transmit the schema first, then data; also, if we have a set of reasonable implementations of this interface, users should hopefully not feel a need to implement it themselves unless they did actually want control over the specifics.

pitrou replied:
Right, but if you look at the RecordBatchStream implementation, it's doing non-trivial stuff with dictionaries.

wesm replied:
I think we'll have to develop some utility code to assist with these details, but it doesn't seem urgent for the moment. If you are creating a custom stream I am not sure right now how to protect the developer from the details of the stream protocol

pitrou replied:
I did the necessary to hide the details in `IpcPayloadWriter`, and the converse is available for reading in `ipc::MessageReader` together with `RecordBatchStreamReader`. There shouldn't be a need to expose this at all.

wesm replied:
OK, how would you like to address this, sequence-wise -- is only the implementation here a problem, or is the public API also a problem? I think it would be much easier to fix both after getting this patch merged since we aren't releasing anytime soon

pitrou replied:
Ok, let's fix this afterwards.

bkietz commented on cpp/src/arrow/flight/server.cc:
```diff
@@ -53,6 +55,9 @@ using ServerWriter = grpc::ServerWriter<T>;
 
 namespace arrow {
+
+using internal::make_unique;
```
accurate; MSVC will occasionally get confused between `internal::make_unique` and `std::make_unique` when it is `using`ed like this. Referring to it with `internal::make_unique` prevents the issue https://issues.apache.org/jira/browse/ARROW-5121

pitrou commented on cpp/src/arrow/flight/server.cc:
```diff
@@ -64,9 +69,11 @@ class FlightMessageReaderImpl : public FlightMessageReader {
  public:
   FlightMessageReaderImpl(const FlightDescriptor& descriptor,
                           std::shared_ptr<Schema> schema,
+                          std::unique_ptr<ipc::DictionaryMemo> dict_memo,
```
See #4319 for a clean, dictionary-compatible, re-implementation of `FlightMessageReader` based on `ipc::MessageReader`.

wesm commented:
I spoke with @romainfrancois and he may not be able to help fix the R bindings until next week, so if it doesn't offend anyone greatly I would add R to allowed failures in Travis CI until R can be fixed

wesm commented:
I addressed the comments so far and added R to `allow_failures` in .travis.yml. I think the only thing left for this to be merge-able is to fix GLib and Ruby -- I will wait for advice from @kou. I also need to merge https://github.com/apache/arrow/pull/3644 and then rebase this on top of it

wesm commented:
Looks like there are still some doxygen issues, and the integration tests are broken (I was hoping I would get lucky there...) so I will address those issues tomorrow (Thursday)

kou commented:
I'll work on this in a few days.
(I'll push some commits to this branch.)

wesm commented:
thanks @kou! I might take a look at the GLib stuff quickly tomorrow to see how involved the changes are, let me know if you start working on it. I need to rebase after ARROW-835 went in so that will take me a little time

kou commented:
OK.
I'll leave a comment when I start committing.

kou commented:
This may be a out of scope topic of this pull request but I want to share.

Can we use`DictionaryType` instead of dictionary type's value type in `DictionaryBuilder`?
In Arrow GLib, we want to detect builder type from `ArrayBuilder` to map corresponding Arrow GLib class: https://github.com/apache/arrow/blob/master/c_glib/arrow-glib/array-builder.cpp#L3970
The current `DictionaryBuilder` only keeps value type. So we can't use `ArrayBuilder::type()` for this.

With this change, can we use `DictionaryType` in `DictionaryBuilder` because we don't need dictionary for `DictionaryType`?

If this is out of scope of this pull request, I'll open a JIRA issue.

pitrou commented:
@wesm I'm currently trying to rebase this, and also fixing CUDA compile failures.

pitrou commented:
I think the integration failure is because the dictionary integration test reuses the same dictionary array for two different fields (with different index types):
```json
{
  "schema": {
    "fields": [
      {
        "name": "dict1_0",
        "type": {
          "name": "utf8"
        },
        "nullable": true,
        "children": [],
        "dictionary": {
          "id": 0,
          "indexType": {
            "name": "int",
            "isSigned": true,
            "bitWidth": 8
          },
          "isOrdered": false
        }
      },
      {
        "name": "dict1_1",
        "type": {
          "name": "utf8"
        },
        "nullable": true,
        "children": [],
        "dictionary": {
          "id": 0,
          "indexType": {
            "name": "int",
            "isSigned": true,
            "bitWidth": 32
          },
          "isOrdered": false
        }
      },
      {
        "name": "dict2_0",
        "type": {
          "name": "int",
          "isSigned": true,
          "bitWidth": 64
        },
        "nullable": true,
        "children": [],
        "dictionary": {
          "id": 1,
          "indexType": {
            "name": "int",
            "isSigned": true,
            "bitWidth": 16
          },
          "isOrdered": false
        }
      }
    ]
  },
  "dictionaries": [
    {
...
```

A complication is with how the dictionary types are serialized into JSON. The "dictionaries" key doesn't allow unserializing the dictionary arrays by themselves, as you have to parse the "schema" key to get the value type...

wesm commented:
@kou the troublesome thing with that is the `DictionaryBuilder` will automatically promote the index type depending on the size of the dictionary, so you might have initialized the builder with `dictionary(int8(), utf8())` but then the result might be `dictionary(int32(), utf8())`. Maybe I'm wrong -- either way I think we can address with a new JIRA

wesm commented:
@pitrou I can look at the integration test failure today -- thank you for rebasing!

pitrou commented:
So the problem is that `DictionaryMemo` expects a `<dictionary field> <-> <dictionary id>` bijection, but that's not true with the JSON integration format.

wesm commented:
Yes, indeed. I'm on it, I will fix.

wesm commented:
OK, I have the integration tests fixed locally. Now multiple fields can reference the same dictionary with no problem. I'll add a unit test for the change I made to `DictionaryMemo` (since the case of multiple fields referencing the same dictionary only occurs in the integration tests)

wesm commented:
Done. I also added rudimentary arguments to toggle the JS and Java integration testers off, it might be worth looking a bit more holistically at integration test CLI options per ARROW-5066

wesm commented:
Integration tests are failing with this error:

```
[dictionary: 0], dict1_1: Int(32, true)[dictionary: 0], dict2_0: Int(16, true)[dictionary: 1]>
Incompatible files
Different schemas:
Schema<dict1_0: Int(8, true)[dictionary: 0], dict1_1: Int(32, true)[dictionary: 1], dict2_0: Int(16, true)[dictionary: 2]>
Schema<dict1_0: Int(8, true)[dictionary: 0], dict1_1: Int(32, true)[dictionary: 0], dict2_0: Int(16, true)[dictionary: 1]>
20:23:31.048 [main] ERROR org.apache.arrow.tools.Integration - Incompatible files
java.lang.IllegalArgumentException: Different schemas:
Schema<dict1_0: Int(8, true)[dictionary: 0], dict1_1: Int(32, true)[dictionary: 1], dict2_0: Int(16, true)[dictionary: 2]>
Schema<dict1_0: Int(8, true)[dictionary: 0], dict1_1: Int(32, true)[dictionary: 0], dict2_0: Int(16, true)[dictionary: 1]>
	at org.apache.arrow.vector.util.Validator.compareSchemas(Validator.java:47)
	at org.apache.arrow.tools.Integration$Command$3.execute(Integration.java:194)
	at org.apache.arrow.tools.Integration.run(Integration.java:118)
	at org.apache.arrow.tools.Integration.main(Integration.java:69)
```

With these changes it is now more difficult to refer to dictionaries multiple times in IPC streams because ids are assigned to fields prior to becoming aware of the dictionaries themselves. I opened ARROW-5340 to spend some time on this -- I'm inclined to remove the multiply-referenced dictionary from the integration tests and leave it for follow up work

kou commented:
@wesm I've created a new issue for https://github.com/apache/arrow/pull/4316#issuecomment-492995395 : https://issues.apache.org/jira/browse/ARROW-5355
So we can ignore this topic in this pull request.

wesm commented:
I removed the multiply-referenced dictionary from the integration tests. I think the dictionary-encoding stuff in Java will need a little bit of work -- it isn't clear to me, for example, why `Field` objects in `Schema` in Java have the dictionary id. In the meantime this will give me some time to sort out how to handle dictionary reuse (which is an optimization basically) in C++

kou commented:
I'll work on this.

kou commented:
Done.

wesm commented:
Thanks @kou! I will rebase this to try to get a green build. @pitrou or @kou can you approve the PR?

Rust build is failing, @sunchao @nevi-me @andygrove do you know what is wrong?

wesm commented:
@xhochy if you are available to peek at this or approve, that would be helpful. I'm happy to address post-merge feedback as well

nevi-me commented:
> Thanks @kou! I will rebase this to try to get a green build. @pitrou or @kou can you approve the PR?
> 
> Rust build is failing, @sunchao @nevi-me @andygrove do you know what is wrong?

Hi @wesm, looks like an issue with a dependency. I'll investigate (https://travis-ci.org/apache/arrow/jobs/533676609#L536)

wesm commented:
I'm inclined to merge this with the Rust build broken since there are a lot of PRs that need to be rebased... if anyone has any objection or wants to look more at the changes please let me know in the next hour or two

nevi-me commented:
> I'm inclined to merge this with the Rust build broken since there are a lot of PRs that need to be rebased... if anyone has any objection or wants to look more at the changes please let me know in the next hour or two

@wesm I'm happy that we don't stop the train for other languages because of the Rust issue. It's only occurring on nightly, and we at least know what the issue is.

wesm commented:
Merging so we can begin rebasing other PRs

wesm commented:
thanks @pitrou and @kou for you help getting this done!
